atlasdb-feign/src/main/java/com/palantir/atlasdb/http/v2/ConjureJavaRuntimeTargetFactory.java

```diff
@@ -0,0 +1,115 @@
+
+import com.google.common.collect.ImmutableList;
+import com.palantir.atlasdb.config.AuxiliaryRemotingParameters;
+import com.palantir.atlasdb.config.ImmutableServerListConfig;
+import com.palantir.atlasdb.config.ServerListConfig;
+import com.palantir.atlasdb.http.AtlasDbRemotingConstants;
+import com.palantir.atlasdb.http.ImmutableInstanceAndVersion;
+import com.palantir.atlasdb.http.PollingRefreshable;
+import com.palantir.atlasdb.http.TargetFactory;
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.conjure.java.client.jaxrs.JaxRsClient;
+import com.palantir.conjure.java.config.ssl.TrustContext;
+import com.palantir.conjure.java.ext.refresh.Refreshable;
+import com.palantir.conjure.java.okhttp.HostMetricsRegistry;
+
+public final class ConjureJavaRuntimeTargetFactory implements TargetFactory {
+    private static final HostMetricsRegistry HOST_METRICS_REGISTRY = new HostMetricsRegistry();
+
+    public static final ConjureJavaRuntimeTargetFactory DEFAULT = new ConjureJavaRuntimeTargetFactory();
+    public static final String CLIENT_VERSION_STRING = "Conjure-Java-Runtime";
+
+    private ConjureJavaRuntimeTargetFactory() {
+        // Use the instances.
+    }
+
+    @Override
+    public <T> InstanceAndVersion<T> createProxy(Optional<TrustContext> trustContext, String uri, Class<T> type,
+            AuxiliaryRemotingParameters parameters) {
+        ClientConfiguration clientConfiguration = ClientOptions.DEFAULT_RETRYING.create(
+                ImmutableList.of(uri),
+                Optional.empty(),
+                trustContext.orElseThrow(() -> new IllegalStateException("CJR requires a trust context")));
+        return wrapWithVersion(JaxRsClient.create(
+                type,
+                addAtlasDbRemotingAgent(parameters.userAgent()),
+                HOST_METRICS_REGISTRY,
+                clientConfiguration));
+    }
+
+    @Override
+    public <T> InstanceAndVersion<T> createProxyWithFailover(
+            ServerListConfig serverListConfig,
+            Class<T> type,
+            AuxiliaryRemotingParameters parameters) {
+        ClientConfiguration clientConfiguration = ClientOptions.DEFAULT_RETRYING.serverListToClient(serverListConfig);
+
+        return wrapWithVersion(JaxRsClient.create(
+                type,
+                addAtlasDbRemotingAgent(parameters.userAgent()),
+                HOST_METRICS_REGISTRY,
+                clientConfiguration));
+    }
+
+    @Override
+    public <T> InstanceAndVersion<T> createLiveReloadingProxyWithFailover(
+            Supplier<ServerListConfig> serverListConfigSupplier,
+            Class<T> type,
+            AuxiliaryRemotingParameters parameters) {
+        Supplier<ServerListConfig> nonEmptyServerList = () -> injectDummyServer(serverListConfigSupplier);
+
+        // TODO (jkong): Thread leak for days. Though no regression over existing implementation in Feign.
```

Comment:
😢

Reply:
why are we leaking threads? plus what happened to the `CachedComposedSupplier`?

Reply:
Yeah, what happened to that?

Reply:
Why: The `PollingRefreshable` has a single threaded executor that needs to be shutdown when the service is closed. This normally won't happen, but can if you close and reopen transaction managers (internal security product used to do this; some use-cases in large internal product also does this). Most services will be fine.

`CachedComposedSupplier`: Yeah, I botched the rebase - will fix.

Reply:
Yeah, I was wondering what happened here. Switching to the CachedComposedSupplier fixes all those issues.

Reply:
The CachedComposedSupplier is still passed to the PollingRefreshable (the change via CCS is that the PollingRefreshable does not do the transform), so I'm confident the thread leak still remains actually.

This is present for legacy reasons; switching Atlas to push-based configuration would be a break of relatively unknown extent (unclear if this is even possible) but remoting seems to demand it - so we have this workaround as a bit of a hack. It is possible to fix this with some proxy tricks (there's possibly a better way), but I would not immediately prioritise that as there is no regression relative to today.